Assets/Scripts/Framework/Statemachine/StateMachine.cs

```diff
@@ -0,0 +1,86 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public abstract class StateMachine : MonoBehaviour
+{
+    [Header("Base StateMachine")]
+    [SerializeField] protected BaseState startingState;
+    [SerializeField] protected BaseState[] states;
+    [SerializeField] protected BaseState currentState;
+
+    [SerializeField] private float removeStateQueueTime;
+    
+    private readonly List<BaseState> _switchStateQueue = new List<BaseState>();
+
+    protected void Awake()
+    {
+        SetStatesParent();
+    }
+
+    private void Start()
+    {
+        EnterStartingState();
+    }
+
+    protected void Update()
+    {
+        currentState.UpdateState(this);
+    }
+    protected void FixedUpdate()
+    {
+        currentState.FixedUpdateState(this);
+
+        if (!currentState.IsValidToSwitch || _switchStateQueue.Count <= 0) return;
+        
+        SwitchState(_switchStateQueue[0]);
+        _switchStateQueue.Remove(_switchStateQueue[0]);
+    }
+
+    private void SetStatesParent()
```

Comment:
Instead of having to go over a list of states to make yourself their parent, why not the other way around? Allow states to make this statemachine their own parent.

I refer you to the Single Responsibility Principle ;)